[Issue]: [rocr-runtime] Compilation with libc++ fails: error: no matching function for call to object of type 'const std::less<void>'

### Problem Description

```
/usr/include/c++/v1/__utility/lazy_synth_three_way_comparator.h:43:54: error: no matching function for call to object of type 'const std::less<void>'
   43 |   _LIBCPP_HIDE_FROM_ABI bool __less() const { return __comp_(__lhs_, __rhs_); }
      |                                                      ^~~~~~~
/usr/include/c++/v1/__tree:1817:20: note: in instantiation of member function 'std::__lazy_compare_result<std::less<void>, hsa_signal_s, hsa_signal_s>::__less' requested here
 1817 |     if (__comp_res.__less()) {
      |                    ^
/usr/include/c++/v1/__tree:1039:38: note: in instantiation of function template specialization 'std::__tree<std::__value_type<hsa_signal_s, long>, std::__map_value_compare<hsa_signal_s, std::pair<const hsa_signal_s, long>, std::less<hsa_signal_t>>, std::allocator<std::pair<const hsa_signal_s, long>>>::__find_equal<hsa_signal_s>' requested here
 1039 |           auto [__parent, __child] = __find_equal(__key);
      |                                      ^
/usr/include/c++/v1/map:1501:18: note: in instantiation of function template specialization 'std::__tree<std::__value_type<hsa_signal_s, long>, std::__map_value_compare<hsa_signal_s, std::pair<const hsa_signal_s, long>, std::less<hsa_signal_t>>, std::allocator<std::pair<const hsa_signal_s, long>>>::__emplace_unique<const std::piecewise_construct_t &, std::tuple<const hsa_signal_s &>, std::tuple<>>' requested here
 1501 |   return __tree_.__emplace_unique(std::piecewise_construct, std::forward_as_tuple(__k), std::forward_as_tuple())
      |                  ^
/var/tmp/portage/dev-libs/rocr-runtime-7.2.0/work/rocr-runtime/runtime/hsa-runtime/core/inc/amd_gpu_agent.h:695:23: note: in instantiation of member function 'std::map<hsa_signal_s, long>::operator[]' requested here
  695 |     scratch_notifiers_[signal] = value;
      |                       ^
/usr/include/c++/v1/__functional/operations.h:374:60: note: candidate template ignored: substitution failure [with _T1 = const hsa_signal_s &, _T2 = const hsa_signal_s &]: invalid operands to binary expression ('const hsa_signal_s' and 'const hsa_signal_s')
  374 |   _LIBCPP_CONSTEXPR_SINCE_CXX14 _LIBCPP_HIDE_FROM_ABI auto operator()(_T1&& __t, _T2&& __u) const
      |                                                            ^
  375 |       noexcept(noexcept(std::forward<_T1>(__t) < std::forward<_T2>(__u))) //
  376 |       -> decltype(std::forward<_T1>(__t) < std::forward<_T2>(__u)) {
      |                                          ~
In file included from /var/tmp/portage/dev-libs/rocr-runtime-7.2.0/work/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:65:
In file included from /var/tmp/portage/dev-libs/rocr-runtime-7.2.0/work/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h:49:
In file included from /usr/include/c++/v1/map:602:
In file included from /usr/include/c++/v1/__tree:42:
/usr/include/c++/v1/__utility/lazy_synth_three_way_comparator.h:44:57: error: no matching function for call to object of type 'const std::less<void>'
   44 |   _LIBCPP_HIDE_FROM_ABI bool __greater() const { return __comp_(__rhs_, __lhs_); }
      |                                                         ^~~~~~~
/usr/include/c++/v1/__tree:1823:27: note: in instantiation of member function 'std::__lazy_compare_result<std::less<void>, hsa_signal_s, hsa_signal_s>::__greater' requested here
 1823 |     } else if (__comp_res.__greater()) {
      |                           ^
/usr/include/c++/v1/__tree:1039:38: note: in instantiation of function template specialization 'std::__tree<std::__value_type<hsa_signal_s, long>, std::__map_value_compare<hsa_signal_s, std::pair<const hsa_signal_s, long>, std::less<hsa_signal_t>>, std::allocator<std::pair<const hsa_signal_s, long>>>::__find_equal<hsa_signal_s>' requested here
 1039 |           auto [__parent, __child] = __find_equal(__key);
      |                                      ^
/usr/include/c++/v1/map:1501:18: note: in instantiation of function template specialization 'std::__tree<std::__value_type<hsa_signal_s, long>, std::__map_value_compare<hsa_signal_s, std::pair<const hsa_signal_s, long>, std::less<hsa_signal_t>>, std::allocator<std::pair<const hsa_signal_s, long>>>::__emplace_unique<const std::piecewise_construct_t &, std::tuple<const hsa_signal_s &>, std::tuple<>>' requested here
 1501 |   return __tree_.__emplace_unique(std::piecewise_construct, std::forward_as_tuple(__k), std::forward_as_tuple())
      |                  ^
/var/tmp/portage/dev-libs/rocr-runtime-7.2.0/work/rocr-runtime/runtime/hsa-runtime/core/inc/amd_gpu_agent.h:695:23: note: in instantiation of member function 'std::map<hsa_signal_s, long>::operator[]' requested here
  695 |     scratch_notifiers_[signal] = value;
      |                       ^
/usr/include/c++/v1/__functional/operations.h:374:60: note: candidate template ignored: substitution failure [with _T1 = const hsa_signal_s &, _T2 = const hsa_signal_s &]: invalid operands to binary expression ('const hsa_signal_s' and 'const hsa_signal_s')
  374 |   _LIBCPP_CONSTEXPR_SINCE_CXX14 _LIBCPP_HIDE_FROM_ABI auto operator()(_T1&& __t, _T2&& __u) const
      |                                                            ^
  375 |       noexcept(noexcept(std::forward<_T1>(__t) < std::forward<_T2>(__u))) //
  376 |       -> decltype(std::forward<_T1>(__t) < std::forward<_T2>(__u)) {
      |                                          ~
8 warnings and 2 errors generated.
```

From what I see, libstdc++ compiles due too loose standard following, allowing to compare some structs without a comparator.

Adding comparator fixes the issue; see attached pull-request.

### Operating System

Gentoo / llvm profile

### CPU

-

### GPU

-

### ROCm Version

ROCm 7.2.0

### ROCm Component

ROCR-Runtime

### Steps to Reproduce

_No response_

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

_No response_

--- a/projects/rocr-runtime/runtime/hsa-runtime/core/inc/amd_gpu_agent.h
+++ b/projects/rocr-runtime/runtime/hsa-runtime/core/inc/amd_gpu_agent.h
@@ -634,8 +634,14 @@ class GpuAgent : public GpuAgentInt {
   // @brief Current short duration scratch memory size.
   size_t scratch_used_large_;
 
+  struct HsaSignalLess {
+    bool operator()(const hsa_signal_t& lhs, const hsa_signal_t& rhs) const {
+      return lhs.handle < rhs.handle;
+    }
+  };
+
   // @brief Notifications for scratch release.
-  std::map<hsa_signal_t, hsa_signal_value_t> scratch_notifiers_;
+  std::map<hsa_signal_t, hsa_signal_value_t, HsaSignalLess> scratch_notifiers_;
 
   // @brief Default scratch size per queue.
   size_t queue_scratch_len_;

